torch/csrc/api/include/torch/nn/utils/clip_grad.h

```diff
@@ -23,43 +29,55 @@ inline double clip_grad_norm_(
       params_with_grad.push_back(param);
     }
   }
-  double total_norm = 0.0;
+
+  if (params_with_grad.empty()) {
+    return 0.0;
+  }
+
+  Tensor total_norm_tensor;
   if (norm_type == std::numeric_limits<double>::infinity()) {
+    std::vector<Tensor> norms;
+    norms.reserve(params_with_grad.size());
+
     for (const auto& param : params_with_grad) {
-      auto param_max = param.grad().data().abs().max().item().toDouble();
-      if (param_max > total_norm || std::isnan(param_max)) {
-        total_norm = param_max;
-      }
+      norms.emplace_back(param.grad().data().abs().max());
     }
+    total_norm_tensor = (norms.size() == 1) ? norms[0] : torch::max(torch::stack(norms));
   } else if (norm_type == 0) {
-    total_norm = static_cast<double>(params_with_grad.size());
+    total_norm_tensor = torch::full({}, static_cast<double>(params_with_grad.size()));
```

Comment:
I'm not all that familiar with the C++ tensor creation APIs, is there a better way of creating a scalar tensor?

And on an unrelated note: order 0 norm is defined in [`torch.linalg.norm`](https://pytorch.org/docs/stable/generated/torch.linalg.norm.html#torch.linalg.norm) as the number of non-zero entries, but in this implementation it's just the number of parameters?

Reply:
`torch::full()` is fine here imo.

Yeah, it does look like there's a discrepancy between here and the order 0 norm definition in `torch.linalg.norm`. I guess there's an implicit assumption that all grads are nonzero here? Regardless, we should leave as-is for backwards compatibility.